addons/l10n_in/models/company.py

```diff
@@ -12,3 +12,17 @@ class ResCompany(models.Model):
         ],
         string="HSN Code Digit",
     )
+
+    def write(self, values):
+        if 'vat' in values:
+            values['l10n_in_hsn_code_digit'] = False if not values['vat'] else '4'
+        elif 'l10n_in_hsn_code_digit' in values and not values['l10n_in_hsn_code_digit']:
+            values['l10n_in_hsn_code_digit'] = '4'
+        return super(ResCompany, self).write(values)
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        companies = super().create(vals_list)
+        for company in companies.filtered(lambda c: c.vat):
+            company.l10n_in_hsn_code_digit = "4"
+        return companies
```

Comment:
compute/onchange of l10n_in_hsn_code_digit instead of write/create.  Won't the government system already throw an error anyways?